kv-storage/src/memory.rs

```diff
@@ -1 +1,153 @@
+use super::*;
+use std::collections::HashMap;
 
+type DB = HashMap<Hash256, Vec<u8>>;
+
+#[derive(Clone)]
+pub struct IMDB {
+    db: DB,
+    checkpoint: DB,
+}
+
+impl IMDB {
+    pub async fn new() -> Self {
+        IMDB {
+            db: DB::new(),
+            checkpoint: DB::new(),
+        }
+    }
+
+    pub async fn open(&self) -> Self {
+        self.clone()
+    }
+}
+
+#[async_trait]
+impl KVStorage for IMDB {
+    async fn commit_checkpoint(&mut self) -> Result<(), Error> {
+        self.checkpoint = self.db.clone();
+
+        Ok(())
+    }
+
+    async fn revert_to_latest_checkpoint(&mut self) -> Result<(), Error> {
+        self.db = self.checkpoint.clone();
+
+        Ok(())
+    }
+
+    async fn insert_or_update(&mut self, key: Hash256, value: &[u8]) -> Result<(), Error> {
+        self.db.insert(key, value.to_vec());
+
+        Ok(())
+    }
+
+    async fn remove(&mut self, key: Hash256) -> Result<(), Error> {
+        match self.db.remove(&key) {
```

Comment:
Let's use https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or for the simplicity

Reply:
Cannot implement using `ok_or` since `HashMap::remove` returns `Option<Vec<u8>>`

Reply:
Fixed.